Packable mode support

tangxifan opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This pull request aims to enhance users' experience in VPR packer.
It allows users to specify if a ``<mode>`` of a ``<pb_type>`` in the architecture XML to be visible/invisible to VPR packer.

#### Description
<!--- Describe your changes in detail -->
Changes are made to ``libarchfpga`` and VPR packer to support the feature.
- New XML syntax ``packable=<bool>`` has been added to ``libarchfpga`` as a property of XML node ``<mode>``. 
- Data structure ``t_mode`` has a new variable of ``packable`` to store the property.
- VPR pack is modified to flag infeasible packing when 
    - Packer tries to consider pack patterns that are defined in an unpackable ``<mode>``
    - Packer tries to map resource to a ``<pb_type>`` under an unpackable ``<mode>``
    - Logic block router tries to expand edges that belong to an unpackable ``<mode>``
- Documentation has been updated for the new syntax

#### Related Issue
<!--- Pull requests should be related to open issues -->

<!--- If suggesting a new feature or change, please discuss it in an issue first -->
This issues has been presented and discussed during the VTR developers' meeting at FPGA'2020.
Currently, it is hard to debug a ``<complexblock>`` with a deep hierarchy, especially due to multi-mode ``<pb_type>``. 
Packer may drop the ball due to a dangling net/connection or other minor issues in the ``<pb_type>``, which VPR's architecture checker will not flag fatal errors.
When packing fails, it is hard for users to spot which ``<mode>`` or which  ``<pb_type>`` causes the problems.

<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Now, users can use XML syntax ``packable`` to specify if a ``<mode>`` of a ``<pb_type>`` in the architecture XML to be visible/invisible to VPR packer.
By default, every ``<mode>`` is packable.

For example, 

    `<mode name="lut6" packable="false">`

When users define an unpackable ``<mode>``, any child ``<pb_type>`` and any mode under the children will be unpackable.

As such, users can turn on/off any ``<mode>`` for packer and quickly spot which ``<mode>`` causes packing to fail.

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
It passed CI regression tests without QoR changes.
Passed artificial XML with unpackable mode definition.

<!--- Include details of your testing environment, and the tests you ran to -->

<!--- see how your change affects other areas of the code, etc. -->
No impact on other codes except packer.

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (change which fixes an issue)
- [ X] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation
- [ X] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ X] All new and existing tests passed

mithro commented:
@tangxifan -- I think this needs a "howto doc" which demonstrates how to debug an issue with the packer and thus explains why this feature exists. It seems like you could just use the verbose packing output?

kmurray commented on doc/src/arch/reference.rst:
```diff
@@ -1454,7 +1462,7 @@ The following tags are common to all <pb_type> tags:
         </pb_type>
 
-    specifies the ``lut`` pb_type can be used as either a single 6-input LUT, or as two 5-input LUTs (but not both).
+    specifies the ``lut`` pb_type can be used as either a single 6-input LUT, or as two 5-input LUTs (but not both). Mode ``lut5x2`` is set to be unpackable, and therefore will not be considered for mapping resources during VPR packing stage. Mode ``lut6`` is by default set to be packable, where VPR packer will try to map resources.
```
I'm not sure putting packable=false in the example is a great idea, since it is really a power-user debugging option and hence not typical usage.

I could anticipate a new user thinking: 'Why would you define this lut5x2 mode but make it unpackable?'. Probably better to leave it out here, and make a packer debugging tutorial like @mithro suggested, which can be cross-refed here (e.g. ``.. seealso::`).

tangxifan replied:
Indeed, this is a feature for power users to debug the packer and architecture file.
I think a tutorial for packer debugging where bugs in architecture file can be quickly spotted by using the ``disable_packing`` switch and the verbose output of packer.

kmurray commented on doc/src/arch/reference.rst:
```diff
@@ -1416,7 +1416,7 @@ The following tags are common to all <pb_type> tags:
     *See above descriptions on inputs*
 
-.. arch:tag:: <mode name="string">
+.. arch:tag:: <mode name="string" packable="bool">
```
Since we expect typical usage to not specify packable, perhaps we should give it a more exceptional name.

Perhaps something like `disable_pack` or `disable_packing`. Not sure

tangxifan replied:
This is a good point. I will change the name.

tangxifan commented:
Hi @mithro @kmurray ,
   I have added a short tutorial about how-to-debug with the ``disable_packing`` switch.
   I notice that actually it can be part of how to build a multi-mode logic block, where we show how to use architecture description language to model a multi-mode logic block and then use the ``disable_packing`` switch to validate it correctness.
   Therefore, I added a tutorial to the architecture part.
   Feel free to advise.

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
sp: conssists -> consists

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
Should probably have a comment here, calling out `num_pb=2` as why this models two 5-LUT + FF pairs

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
Similarly, it would be helpful to have a comment noting `num_pb="1"` here.

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
Probably should clarify that this is being done temporarily to validate the correctness of the mode definitions.

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
'will consider' -> 'will only consider'

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
'functional correct' -> 'being successfully used by the packer'

Functional correctness is a somewhat stronger statement than what is being tested here (e.g. could be interpreted as being formally equivalent to the hardware). It is likely better to just say the packer was able to use it correctly.

kmurray commented on doc/src/arch/reference.rst:
```diff
@@ -1430,6 +1430,16 @@ The following tags are common to all <pb_type> tags:
 
+    :opt_param disable_packing:
+        Specify if a mode is disabled or not for VPR packer.
+        When a mode is defined to be disabled for packing (``disable_packing="true"``), packer will not map any logics to the mode.
```
'logics' -> 'logic'

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
'logics' -> 'logic'

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
Same as above

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
'Finally, after having validated' ... 'that both modes are being successfully used by the packer.'

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
'enable' -> 're-enable'

kmurray commented on doc/src/tutorials/arch/multi_mode_ble.rst:
'Now, VPR's packer will try to choose the best operating modes to use.

mithro commented:
@litghost Could you do a quick look see here?

tangxifan commented:
@kmurray Thanks for the comments. It helps a lot in improving the quality of this tutorial. I have applied these modifications and added the test case ``strong_pack_disable`` to the strong regression tests. I used the same architecture and benchmark as the tutorial.

litghost commented:
I'll take a look at this next week.

tangxifan commented:
Hi all, can I have your latest opinion? It has been quite a while since the last update.

kmurray commented:
Looks good to me.

mithro commented:
@litghost is out of office until Friday, so I think we can proceed with @kmurray's approval.
